File: Java/1022. Sum of Root To Leaf Binary Numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

 // Iterative Solution
class Solution {
    public int sumRootToLeaf(TreeNode root) {
        int total = 0;
        Deque<Pair<TreeNode, Integer>> stack = new ArrayDeque();
        stack.push(new Pair(root, 0));
        
        while(!stack.isEmpty()){
            Pair<TreeNode, Integer> cur = stack.pop();
            root = cur.getKey();
            int path = cur.getValue();
            
            if(root != null){
                path = path*2 + root.val;
                
                // if current node is a leaf node
                if(root.left == null && root.right == null){
                    total += path;
                }else{
                    stack.push(new Pair(root.right, path));
                    stack.push(new Pair(root.left, path));
                }
            }
        }
        
        return total;
    }
}

File: Python3/1022. Sum of Root To Leaf Binary Numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Recursive Solution
class Solution:
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        self.total = 0
        
        def dfs(node, path):
            # if node is null
            if not node:
                return
            
            # update the path sum from top to this node
            path = path*2 + node.val
            
            # if node is a leaf node
            if not node.left and not node.right:
                self.total += path
                
            dfs(node.left, path)
            dfs(node.right, path)
            
        dfs(root, 0)
        return self.total